src/main/java/com/googlecode/lanterna/gui2/ScrollPanel.java

```diff
@@ -0,0 +1,172 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * This scroll panel can be used instead of allowing the TextBox, ActionListBox, CheckBoxList, RadioBoxList components to draw their own ScrollBar.
+ *
+ * This is backwards compatible via this approach:
+ *   1) the scrollable components are added to an instance of one of these ScrollPanels
+ *   2) that causes a flag to be set on them, which their renderers use to not paint the ScrollBar
+ *   3) the ScrollPanel is added to the Container instead of the scrollable components directly
+ * Note: there is a gap in compatability in that custom Renderers created by users of the library may be doing stuff with their ScrollBars (very unlikely)
+ * 
+ * The contained Component is not replaceable, user is advised to discard the instance of ScrollPanel and recreate for that functionality.
+ * 
+ * Note: only extending Panel due to ease of development, since don't know the framework enough to just implement the Component interface
+ *       user should not expect to invoke any of Panel container style method
+ 
+ * TODO: implement Component, don't extends Panel
+ *
+ * @author ginkoblongata
+ */
+public class ScrollPanel extends Panel {
+    private final ScrollableBox scrollableBox;
+    private final ScrollBar verticalScrollBar;
+    private final ScrollBar horizontalScrollBar;
+    
+    private final ScrollPanelLayoutManager scrollPanelLayoutManager;
+
+    /**
+     * Default constructor, creates a new panel with no child components and by default set to a vertical
+     * {@code LinearLayout} layout manager.
+     */
+    public ScrollPanel(ScrollableBox scrollableBox) {
+        this.scrollableBox = scrollableBox;
+        verticalScrollBar = new ScrollBar(Direction.VERTICAL, this);
+        horizontalScrollBar = new ScrollBar(Direction.HORIZONTAL, this);
+        scrollPanelLayoutManager = new ScrollPanelLayoutManager();
+        setLayoutManager(scrollPanelLayoutManager);
+        
+        scrollableBox.setIsWithinScrollPanel(true);
+        addComponent(scrollableBox, Location.CENTER);
+    }
+    
+    public void doPage(boolean isVertical, boolean isLess) {
+        // delegate, don't want the renderer having reference to scrollableBox
+        scrollableBox.doPage(isVertical, isLess);
+    }
+    
+    public void doScroll(boolean isVertical, boolean isLess) {
+        // delegate, don't want the renderer having reference to scrollableBox
+        scrollableBox.doScroll(isVertical, isLess);
+    }
+    
+    public void thumbMouseDown(boolean isVertical, TerminalPosition position) {
+        scrollableBox.thumbMouseDown(isVertical, position);
+    }
+    
+    public void thumbMouseDrag(boolean isVertical, TerminalPosition position) {
+        scrollableBox.thumbMouseDrag(isVertical, position);
+    }
+    public void mouseUp() {
+        scrollableBox.mouseUp();
+    }
+    
+    @Override
+    public void invalidate() {
```

Comment:
Don't think you need to override this, it should work anyway. The panel is marked as invalid in the Panel class, which forces a re-render of the whole thing.

Reply:
Ok, yep works without this method, using existing implementation.